docs/Planner.md

```diff
@@ -1,63 +1,86 @@
+A node receives data directly from its predecessors.
+A node feeds data directly to its successors.
 
-The leaves of the plan DAG represent sources of data and the data flows from bottom up through the tree.
-Again this is inverted from the Query DAG where data flows top down.
+## Planning
+-----------
 
-## Data Frames
+There are two types of query plans - logical and physical.
+The nodes of a logical plan represent logical operations.
+The nodes of a physical plan represent physical operations.
+Logical operations define how data will be transformed.
+Physical operations define how data transformations are implemented.
 
-Data frames are a set of data and their lineage is known.
-Meaning it is known what parent data frames and operations are needed to construct the data frame.
-Using this concept of lineage allows a data frame to be reconstructed if it is loss due to node failure or if its parent data frames are modified.
+A Flux query spec is translated directly into a logical plan.
 
-### Windowing
+## Optimization
+---------------
 
-Data frames will specify their windowing properties. ????
+Optimizing a Flux query occurs in two phases - logical and physical.
 
-## Operations
+During logical optimization, the planner converts an initial logical plan into an equivalent logical plan that the query engine will be able to execute more efficiently. This is achieved via a set of rewrite rules that the planner enacts to transform a plan whenever it encounters an eligible node.
 
-Operations are a definition of a transformation to be applied on one data frame resulting in another.
+During physical optimization, a logical plan is converted into an equivalent physical plan. Physical plan nodes have a cost of execution associated with them. Hence the goal during this phase is to compute the equivalent physical plan with the smallest cost.
 
-### Narrow vs Wide
+## Rewrite Rules and Patterns
+-----------------------------
 
-Operations are classified as either narrow or wide:
+```go
+type Rule interface {
+	// The name of this rule (must be unique)
+	Name() string
 
-* Narrow operations map each parent data frame to exactly one child data frame.
-    Specifically a narrow operation is a one-to-one mapping of parent to child data frames.
-* Wide operations map multiple parent data frames to multiple child data frames.
-    Specifically a wide operation is a many-to-many mapping of parent to child data frames.
+	// Pattern for this rule to match against
+	Pattern() Pattern
 
-This distinction is necessary to precisely define the lineage of a data frame.
+	// Rewrite an operation into an equivalent one
+	Rewrite(PlanNode) (PlanNode, bool)
+}
+```
 
+A rewrite rule defines an equivalence class of queries.
+That is, for any data instance, a query expressed before and after a rewrite rule is applied will always define the same data transformation.
+A rule must match a specified pattern in order to rewrite an operation.
```

Comment:
An example would be nice here.

I think we'll want a separate document that goes into more depth on writing rules, since open source contributors may want to write them.